### PR TITLE
document "advanced" tool test debugging

### PR DESCRIPTION
**Pull request by bernt-matthias**

### CODE REVIEW

File: docs/writing_how_do_i.rst [hunk 1]
```diff
@@ -327,3 +327,60 @@ file.
 .. _Docker Hub: https://hub.docker.com/
 .. _Planemo: http://planemo.readthedocs.org/
 .. _Picard: http://broadinstitute.github.io/picard/
+
+----------------------------------------------------------------------------
+\.\.\. test tools against a package or container in a bioconda pull request?
+----------------------------------------------------------------------------
+
+First, obtain the artifacts of the PR by sending the following message
```
Comment (nsoranzo) [inline docs/writing_how_do_i.rst:335]:
```suggestion
First, obtain the artifacts of the PR by adding this comment:
```

File: docs/writing_how_do_i.rst [hunk 2]
```diff
@@ -327,3 +327,60 @@ file.
...
+A planemo test will then simply use this image::
+
+     $ planemo test ... --biocontainers --no_conda_auto_init ...
+
+--------------------------------------
+\.\.\. interactively debug tool tests?
+--------------------------------------
+
+It can be desirable to interactively debug a tool test. In order to do so, start ``planemo test``
+with the option ``--no_cleanup``. Inspect the output: After starting galaxy tests start. At the
```
Comment (nsoranzo) [inline docs/writing_how_do_i.rst:356]:
```suggestion
with the option ``--no_cleanup``. Inspect the output: After Galaxy starts up, the tests commence. At the
```

File: docs/writing_how_do_i.rst [hunk 3]
```diff
@@ -327,3 +327,60 @@ file.
...
+start of each test one finds a message: ``( TOOL_ID ) > Test-X``. After a some upload jobs the
```
Comment (nsoranzo) [inline docs/writing_how_do_i.rst:357]:
```suggestion
start of each test one finds a message: ``( <TOOL_ID> ) > Test-N``. After some upload jobs, the
```

File: docs/writing_how_do_i.rst [hunk 4]
```diff
@@ -327,3 +327,60 @@ file.
...
+actual tool job is started (its the last before the next test is executed). There you will find
```
Comment (nsoranzo) [inline docs/writing_how_do_i.rst:358]:
```suggestion
actual tool job is started (it is the last before the next test is executed). There you will find
```

File: docs/writing_how_do_i.rst [hunk 5]
```diff
@@ -327,3 +327,60 @@ file.
...
+a message like ``Built script [/tmp/tmp1zixgse3/job_working_directory/000/3/tool_script.sh]``
+
+In this case ``/tmp/tmp1zixgse3/job_working_directory/000/3/`` is the job dir. It contains some
+files and directories of interest: 
+
+- ``tool_script.sh`` the bash script generated from the tool's ``command`` and ``version_command``
+  tags plus some boiler plate code
+- ``galaxy_3.sh`` (note that the number may be different) a schell script setting up the environment
```
Comment (nsoranzo) [inline docs/writing_how_do_i.rst:366]:
```suggestion
- ``galaxy_3.sh`` (note that the number may be different): a shell script setting up the environment
```

File: docs/writing_how_do_i.rst [hunk 6]
```diff
@@ -327,3 +327,60 @@ file.
...
+It can be desirable to interactively debug a tool test. In order to do so, start ``planemo test``
+with the option ``--no_cleanup``. Inspect the output: After starting galaxy tests start. At the
+start of each test one finds a message: ``( TOOL_ID ) > Test-X``. After a some upload jobs the
+actual tool job is started (its the last before the next test is executed). There you will find
+a message like ``Built script [/tmp/tmp1zixgse3/job_working_directory/000/3/tool_script.sh]``
+
+In this case ``/tmp/tmp1zixgse3/job_working_directory/000/3/`` is the job dir. It contains some
+files and directories of interest: 
+
+- ``tool_script.sh`` the bash script generated from the tool's ``command`` and ``version_command``
```
Comment (nsoranzo) [inline docs/writing_how_do_i.rst:364]:
```suggestion
- ``tool_script.sh``: the bash script generated from the tool's ``command`` and ``version_command``
```

File: docs/writing_how_do_i.rst [hunk 7]
```diff
@@ -327,3 +327,60 @@ file.
...
+  tags plus some boiler plate code
+- ``galaxy_3.sh`` (note that the number may be different) a schell script setting up the environment
+  (e.g. paths and environment variables), starting the ``tool_script.sh``, and postprocessing
+  (e.g. error handling and setting metadata)
+- ``working`` the jobs working dir
+- ``outputs`` a dir containing the jobs stderr and stdout
```
Comment (nsoranzo) [inline docs/writing_how_do_i.rst:370]:
```suggestion
- ``working``: the job working directory
- ``outputs``: a directory containing the job stderr and stdout
```

File: docs/writing_how_do_i.rst [hunk 8]
```diff
@@ -327,3 +327,60 @@ file.
...
+
+For a tool test that uses a conda environment to resolve the requirements one can simply change
+into ``working`` and execute ``../tool_script.sh`` (works as long as no special environment variables
+are used; in this case ``../galaxy_3.sh`` needs to be executed after cleaning the job dir). 
+By editing the tool script one may understand/fix problems in the ``command`` block faster than by
+rerunning ``planemo test`` over and over again.
+
+Alternatively one can change into the ``working`` dir and load the conda environment
+(the code to do so can be found in ``tool_script.sh``: ``. PATH_TO_CONDA_ENV activate``). 
+Afterwards one can execute individual commands, e.g. those found in ``tool_script.sh`` or varaints.
```
Comment (nsoranzo) [inline docs/writing_how_do_i.rst:380]:
```suggestion
Afterwards one can execute individual commands, e.g. those found in ``tool_script.sh`` or variants.
```

File: docs/writing_how_do_i.rst [hunk 9]
```diff
@@ -327,3 +327,60 @@ file.
...
+
+For a tool test that uses a docker to to resolve the requirements one needs to execute
```
Comment (nsoranzo) [inline docs/writing_how_do_i.rst:382]:
```suggestion
For a tool test that uses Docker to to resolve the requirements one needs to execute 
```

**Review by nsoranzo — APPROVED** [review]

**Comment by jmchilton** [comment]
I missed this when opened but very cool, really awesome stuff! Thanks so much.

**Pull request merged**